File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.cc
#include <iostream>
#include <map>
#include <vector>


using namespace std;
int N;
long long B;
vector<vector<int>> originMap;
map<long long, vector<vector<int>>> maps;
vector<vector<int>> conquer(long long Num);
vector<vector<int>> sub(long long A, long long B);

vector<vector<int>> conquer(long long Num)
{
	if (maps.find(Num) != maps.end())
	{
		return maps[Num];
	}
	return sub(Num / 2l, Num % 2l == 1l ? Num / 2l + 1l : Num / 2);
}

vector<vector<int>> sub(long long A, long long B)
{
	vector<vector<int>> tmp1 = conquer(A);
	vector<vector<int>> tmp2 = conquer(B);
	vector<vector<int>> response;

	for (int i = 0;i < N;i++)
	{
		response.push_back({});
		for (int j = 0;j < N;j++)
		{
			int sum = 0;
			for (int k = 0;k < N;k++)
			{
				sum += tmp1[i][k] * tmp2[k][j];
			}
			response[i].push_back(sum % 1000);
		}
	}
	maps.insert({ A+B, response });
	return response;
}

int main(void)
{
	cin >> N >> B;
	

	for (int i = 0;i < N;i++)
	{
		originMap.push_back({});
		int num;
		for (int j = 0;j < N;j++)
		{
			cin >> num;
			originMap[i].push_back(num%1000);
		}
	}

	maps.insert({1, originMap});
	vector<vector<int>> answer = conquer(B);
		
	for (int i = 0;i < N;i++)
	{
		for (int j = 0;j < N;j++)
		{
			cout << answer[i][j] << " ";
		}
		cout << endl;
	}

	return 0;
}

File: 백준/Gold/10830. 행렬 제곱/행렬 제곱.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.HashMap;
import java.util.StringTokenizer;

public class Main {
	static HashMap<Long, int[][]> maps;
	static int[][] originMap;
	static int N;
	static long B;
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		B = Long.parseLong(st.nextToken());
		
		originMap = new int[N][N];
		maps = new HashMap<>();
		for(int i=0;i<N;i++)
		{
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<N;j++)
			{
				originMap[i][j] = Integer.parseInt(st.nextToken())%1000;
			}
		}
		
		maps.put(1l, originMap);
		int[][] answer =conquer(B);
		
		for(int i=0;i<N;i++)
		{
			for(int j=0;j<N;j++)
			{
				System.out.print(answer[i][j]+" ");
			}
			System.out.println();
		}
		
	}

	private static int[][] conquer(long num) {
		
		if(maps.containsKey(num)) return maps.get(num); 
		else return sub(num/2l, num%2l==1l ? num/2l+1l : num/2);
		
	}

	private static int[][] sub(long l, long m) {
		int[][] tmp1 = conquer(l);
		int[][] tmp2 = conquer(m);
		int[][] res = new int[N][N];
		
		for(int i=0;i<N;i++)
		{
			for(int j=0;j<N;j++)
			{
				int sum = 0;
				for(int k=0;k<N;k++)
				{
					sum += tmp1[i][k] * tmp2[k][j];
				}
				res[i][j] = sum%1000;
			}
		}
		
		maps.put(l+m, res);
		return res;
		
	}

}